miio/push_server.py

```diff
@@ -0,0 +1,401 @@
+        """Start Miio push server."""
+        if self._listen_couroutine is not None:
+            _LOGGER.error("Miio push server already started, not starting another one.")
+            return
+
+        listen_task = self._create_udp_server()
+        _, self._listen_couroutine = await listen_task
+
+    def stop(self):
+        """Stop Miio push server."""
+        if self._listen_couroutine is None:
+            return
+
+        for ip in list(self._registered_devices):
+            self.unregister_miio_device(self._registered_devices[ip]["device"])
+
+        self._listen_couroutine.close()
+        self._listen_couroutine = None
+
+    def subscribe_event(self, device: Device, event_info: EventInfo):
+        """Subscribe to a event such that the device will start pushing data for that
+        event."""
+        if device.ip not in self._registered_devices:
+            _LOGGER.error("Can not subscribe event, miio device not yet registered")
+            return None
+
+        if self.server_ip is None:
+            _LOGGER.error("Can not subscribe event withouth starting the push server")
+            return None
+
+        self._event_id = self._event_id + 1
+        event_id = f"x.scene.{self._event_id}"
+
+        event_payload = self._construct_event(event_id, event_info, device)
+
+        response = device.send(
+            "send_data_frame",
+            {
+                "cur": 0,
+                "data": event_payload,
+                "data_tkn": 29576,
+                "total": 1,
+                "type": "scene",
+            },
+        )
+
+        if response != ["ok"]:
+            _LOGGER.error(
+                "Error subscribing event, response %s, event_payload %s",
+                response,
+                event_payload,
+            )
+            return None
+
+        event_ids = self._registered_devices[device.ip]["event_ids"]
+        event_ids.append(event_id)
+
+        return event_id
+
+    def unsubscribe_event(self, device: Device, event_id):
+        """Unsubscribe from a event by id."""
+        result = device.send("miIO.xdel", [event_id])
+        if result == ["ok"]:
+            event_ids = self._registered_devices[device.ip]["event_ids"]
+            if event_id in event_ids:
+                event_ids.remove(event_id)
+        else:
+            _LOGGER.error("Error removing event_id %s: %s", event_id, result)
+
+        return result
+
+    def _construct_event(  # nosec
+        self,
+        event_id,
+        info: EventInfo,
+        device: Device,
+    ):
+        """Construct the event data payload needed to subscribe to an event."""
+        if info.event is None:
+            info.event = info.action
+        if info.source_sid is None:
+            info.source_sid = str(device.device_id)
+        if info.source_model is None:
+            info.source_model = device.model
+        if info.source_token is None:
+            info.source_token = device.token
```

Comment:
As in the test script EventInfo is constructed like:
  event_info = EventInfo(
      action="click_ch0",
      extra="[1,13,1,85,[0,1],0,0]",
      source_sid=button.sid,
      source_model=button.zigbee_model,
      source_token=gateway.token,
  )

Reply:
Okay, this can be cleaned up afterwards, it would be better to pass the source device as-is and read the required information out from it :-)

Reply:
That is done for the alarm event because that event actually belongs to the wifi-device (in this case the gateway).
This expample showed a zigbee devices which is a bit more difficult since the Device passed is the gateway (for communication) but the SID and model are from the zigbee devices so there are explicitly specified.

For wifi-devices you do not have to specify the source properties and they are indeed taken from the device passed as a whole:
```

event_info = EventInfo(
        action="alarm_triggering",
        extra="[1,19,1,111,[0,1],2,0]",
        trigger_token=gateway.token,
  )

push_server.subscribe_event(gateway, event_info)
```